### PR TITLE
Rasterdraw

### PR DESCRIPTION
**Pull request by DennisZemela**
Added the rasterDraw() and a recording for the flowDraw() in order to compare  the error signal from the eye encoders with the mean optical flow

The changes in the decodeEvents and vTrack.h are not mine

### CODE REVIEW

File: python/decodeEvents/decode_events.py [hunk 1]
```diff
@@ -152,5 +152,5 @@ def load_SkinEvents_from_yarp(self, AE_file_path):
 if __name__ == '__main__':
 
     dm = DataManager()
-    dm.load_AE_from_yarp('acquisitions/circle10042019')
+    dm.load_AE_from_yarp('data_from_dumper')
```
Comment (arrenglover) [inline python/decodeEvents/decode_events.py]: don't add anything outside the scope of the raster draw to the pull request

File: src/applications/gazeDemo/include/vTrackToRobot.h [hunk 2]
```diff
@@ -295,6 +295,13 @@ class eyeControlPID
         velocity[4]=eyes_pan;           // eyes pan
         velocity[5]=eyes_ver;           // eyes vergence
         ivel->velocityMove(velocity.data());
+
+        yarp::sig::Vector accls(6, 0.0);
```
Comment (arrenglover) [inline src/applications/gazeDemo/include/vTrackToRobot.h]: don't add anything outside the scope of the raster draw to the pull request

File: src/applications/gazeDemo/include/vTrackToRobot.h [hunk 3]
```diff
@@ -328,6 +335,13 @@ class eyeControlPID
         velocity[4]=eyes_pan;           // eyes pan
         velocity[5]=eyes_ver;           // eyes vergence
         ivel->velocityMove(velocity.data());
+
```
Comment (arrenglover) [inline src/applications/gazeDemo/include/vTrackToRobot.h]: don't add anything outside the scope of the raster draw to the pull request

File: src/processing/vCluster/src/eventClustering.cpp [hunk 4]
```diff
@@ -69,8 +69,7 @@ bool EventClustering::configure(yarp::os::ResourceFinder &rf)
     double clusterLimit =
             rf.check("clusterLimit", yarp::os::Value(-1)).asDouble();
 
-
-
+    yWarning() << Tact;
```
Comment (arrenglover) [inline src/processing/vCluster/src/eventClustering.cpp]: don't add anything outside the scope of the raster draw to the pull request

File: src/processing/vCluster/src/eventClustering.cpp [hunk 5]
```diff
@@ -208,7 +207,7 @@ void EventBottleManager::onRead(ev::vBottle &bot)
         auto aep = ev::as_event<ev::AddressEvent>(*qi);
         if(!aep) continue;
 
-        short channel           = aep->getChannel();
+        short channel = aep->getChannel();
```
Comment (arrenglover) [inline src/processing/vCluster/src/eventClustering.cpp]: don't add anything outside the scope of the raster draw to the pull request

File: src/processing/vFramer/include/vDraw.h [hunk 6]
```diff
@@ -51,10 +51,11 @@ class vDraw {
     unsigned int display_window;
     unsigned int max_window;
     bool flip;
+    bool scaling;
 
 public:
 
-    vDraw() : Xlimit(304), Ylimit(240), flip(false)
+    vDraw() : Xlimit(304), Ylimit(240), flip(true)
```
Comment (arrenglover) [inline src/processing/vFramer/include/vDraw.h]: don't change the default options of the base class vDraw - it will affect all other drawers too. The defualt is FALSE

File: src/processing/vFramer/include/vDraw.h [hunk 7]
```diff
@@ -125,6 +131,34 @@ class accDraw : public vDraw {
 
 };
 
+class rasterDraw : public vDraw {
+
+protected:
+    #define XLIMIT 304          //array needs constants
```
Comment (arrenglover) [inline src/processing/vFramer/include/vDraw.h]: The raster drawer won't work with hard coded resolutions. We have to be flexible with resolutions

File: src/processing/vFramer/include/vDraw.h [hunk 8]
```diff
@@ -125,6 +131,34 @@ class accDraw : public vDraw {
...
+    #define YLIMIT 240
+
+    double pixelStorage[YLIMIT][XLIMIT];
```
Comment (arrenglover) [inline src/processing/vFramer/include/vDraw.h]: there should be no need to add extra storage. plot directly from the events given to you

File: src/processing/vFramer/include/vDraw.h [hunk 9]
```diff
@@ -51,10 +51,11 @@ class vDraw {
     unsigned int display_window;
     unsigned int max_window;
     bool flip;
+    bool scaling;
```
Comment (arrenglover) [inline src/processing/vFramer/include/vDraw.h]: I don't think the addition of a scale parameter is the best approach. It would be better to automatically scale the y-axis by the largest neuron id you have read so far.

File: src/processing/vFramer/include/vDraw.h [hunk 10]
```diff
@@ -125,6 +131,34 @@ class accDraw : public vDraw {
...
 
+class rasterDraw : public vDraw {
+
+protected:
+    #define XLIMIT 304          //array needs constants
+    #define YLIMIT 240
+
+    double pixelStorage[YLIMIT][XLIMIT];
+    int Xlimit = XLIMIT;
+    double Ylimit = YLIMIT;
```
Comment (arrenglover) [inline src/processing/vFramer/include/vDraw.h]: Xlimit and Ylimit are provided in the base-class and should not be replicated here

File: src/processing/vFramer/include/vDraw.h [hunk 11]
```diff
@@ -171,8 +205,27 @@ class skinsampleDraw : public vDraw {
 
 class flowDraw : public vDraw {
 
+protected:
+    int Xlimit;
```
Comment (arrenglover) [inline src/processing/vFramer/include/vDraw.h]: this is outside the scope of the vDraw - and is also doing something incorrectly

File: src/processing/vFramer/src/other_drawers.cpp [hunk 12]
```diff
@@ -268,3 +268,99 @@ void overlayStereoDraw::draw(cv::Mat &image, const ev::vQueue &eSet, int vTime)
...
+    // check if eSet contains elements
+    if(eSet.empty()){
+        return;
+    }
+    // get latest timestamp
+    if(vTime < 0){
+        vTime = eSet.back()->stamp;
+    }
+    // Reverse Timeelement-Iterator:
+    for(int y=(Ylimit-1.0); y>=0; y--){
```
Comment (arrenglover) [inline src/processing/vFramer/src/other_drawers.cpp]: there should be no need for pixel storage at all. It should be a matter of taking all the events in eSet and plotting them on the image as you like. I don't think the implementation is one we want to use.

**Comment by marco-monforte** [comment]
Yeah, I can take care of it with the help of @DennisZemela for the rasterDrawer. I will have a better look at the code asap.

**Comment by arrenglover** [comment]
there is a fair amount of changes to be made in both the Pull Request and the Rasterdrawer itself. If you want I can find time to have a look at how it could be done with you @marco-monforte - if you are still working on the VOR project, otherwise we can do it with whoever is next working on VOR.

**Comment by marco-monforte** [comment]
Well, yes, it would be much better. I'm still on the project and probably Ander is going to help me with it. We could plan some time in the next week to work on this the three of us.
